org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/LanguageServerImpl.xtend

```diff
@@ -411,11 +433,33 @@ import org.eclipse.xtext.validation.Issue
 	}
 
 	override codeLens(CodeLensParams params) {
-		throw new UnsupportedOperationException("TODO: auto-generated method stub")
+		return requestManager.runRead[ cancelIndicator |
+			val uri = params.textDocument.uri.toUri
+			val resourceServiceProvider = uri.resourceServiceProvider
+			val codeLensService = resourceServiceProvider?.get(ICodeLensService)
+			if (codeLensService === null)
+				return emptyList
+
+			return workspaceManager.doRead(uri) [ document, resource |
+				codeLensService.computeCodeLenses(document, resource, params, cancelIndicator)
+			]
+		]
 	}
 
 	override resolveCodeLens(CodeLens unresolved) {
-		return CompletableFuture.completedFuture(unresolved)
+		return CompletableFutures.computeAsync[ cancelChecker |
+			var CodeLens current = unresolved
+			for (language : this.allLanguages) {
+				val codeLensService = language.get(ICodeLensResolver)
+				if (codeLensService !== null) {
+					current = codeLensService.resolveCodeLens(this.access, current, [ 
```

Comment:
- the first indicator from `resolveCodeLens` to cancel JSON-RPC request that is passed as a 3rd argument
- the second from `access.doRead` to cancel the read request

With `doRead` it would be the same, it is because `access.doRead` starts a new read request. But with `doRead` there is no need to use `access` one can directly read documents.